File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.py
from collections import deque

N, M, V = map(int, input().split())
graph = [[False] * (N + 1) for _ in range(N + 1)]
for i in range(M):
    a,b= map(int, input().split())
    graph[a][b] = True
    graph[b][a] = True
    
visited1 = [False]*(N+1)
visited2 = [False]*(N+1)

def dfs(V):
    visited1[V] = True
    print(V, end=" ")
    for i in range(1, N + 1):
        if not visited1[i] and graph[V][i]:
            dfs(i)
def bfs(V):
    queue = deque([V])
    visited2[V]=True
    while queue:
        V = queue.popleft()
        print(V, end=" ")
        for i in range(N+1):
            if not visited2[i] and graph[V][i]:
                queue.append(i)
                visited2[i]= True
    
dfs(V)
print()
bfs(V)
    

File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.js
const input = require("fs").readFileSync("/dev/stdin").toString().trim().split('\n')

let [n, m, v] = input.shift().split(" ").map(Number);
let graph = [...Array(n + 1)].map((e) => []);

for (let i = 0; i < m; i++) {
  let [start, end] = input[i].split(" ").map(Number);
  graph[start].push(end);
  graph[end].push(start);
}

const dfs = (graph, start) => {
    const visited=[]
    let needVisit = []
    needVisit.push(start)
    while(needVisit.length){
        const node = needVisit.pop() // stack
        if(!visited.includes(node)){
            visited.push(node)
            let nodes = graph[node]
            needVisit = [...needVisit, ...nodes.sort((a,b)=>b-a)] // 스택 때문에 내림차순
        }
    }
    return visited
}

const bfs = (graph, start) =>{
    const visited=[]
    let needVisit = []
    needVisit.push(start)
    while(needVisit.length){
        const node = needVisit.shift() // queue
        if(!visited.includes(node)){
            visited.push(node)
            let nodes = graph[node]
            needVisit = [...needVisit, ...nodes.sort((a,b)=> a-b)]
        }
    }
    return visited
}

console.log(dfs(graph,v).join(' '))
console.log(bfs(graph,v).join(' '))
